controllers/drpolicy_controller.go

```diff
@@ -0,0 +1,103 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ramendrv1alpha1 "github.com/ramendr/ramen/api/v1alpha1"
+	"github.com/ramendr/ramen/controllers/util"
+)
+
+// DRPolicyReconciler reconciles a DRPolicy object
+type DRPolicyReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=drpolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=drpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=drpolicies/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the DRPolicy object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("DRPolicy", req.NamespacedName)
+	log.Info("reconcile enter")
+
+	defer log.Info("reconcile exit")
+
+	drpolicy := &ramendrv1alpha1.DRPolicy{}
+	if err := r.Client.Get(ctx, req.NamespacedName, drpolicy); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("get: %w", err))
+	}
+
+	manifestWorkUtil := util.MWUtil{Client: r.Client, Ctx: ctx, Log: log, InstName: "", InstNamespace: ""}
+
+	const finalizerName = "drpolicies.ramendr.openshift.io/ramen"
+
+	switch drpolicy.ObjectMeta.DeletionTimestamp.IsZero() {
+	case true:
+		log.Info("create/update")
+
+		controllerutil.AddFinalizer(drpolicy, finalizerName)
+
+		if err := r.Update(ctx, drpolicy); err != nil {
+			return ctrl.Result{}, fmt.Errorf("finalizer add update: %w", err)
+		}
+
+		if err := manifestWorkUtil.ClusterRolesCreate(drpolicy); err != nil {
+			return ctrl.Result{}, fmt.Errorf("cluster roles create: %w", err)
+		}
+	default:
+		log.Info("delete")
+
+		if err := manifestWorkUtil.ClusterRolesDelete(drpolicy); err != nil {
+			return ctrl.Result{}, fmt.Errorf("cluster roles delete: %w", err)
+		}
+
+		controllerutil.RemoveFinalizer(drpolicy, finalizerName)
```

Comment:
same here

Reply:
`RemoveFinalizer` [checks if it exists](https://github.com/kubernetes-sigs/controller-runtime/blob/3c54acbad091d40ef80d467943c5d7126c4b8291/pkg/controller/controllerutil/controllerutil.go#L366) before removing it

Reply:
Same as above

Reply:
Optimization to update only if finalizer removed has been added.